added function parameter for secure cookies, added test, added api ex…

ry8806 opened the pull request:
added function parameter in localstorageservice.cookie.set() for secure cookies client side using the "secure" attribute mentioned in the W3C CSP Headers Spec (server side): https://www.w3.org/TR/csp-cookies/#grammardef-secure

added tests and fixed some of the README.md

ry8806 commented:
#272

ealves-pt commented:
@ry8806 what about the config phase? For example, If by default I want to set all the cookies to be stored as secure?

ealves-pt commented:
This PR will close this old PR, #185

ealves-pt commented:
@ry8806 when you can rebase with the master and update the PR.

ealves-pt commented:
This PR also closes #250

ry8806 commented:
@ealves-pt I'll get on this and implement the config phase, I'll aim to do it within 24hrs

ealves-pt commented:
:+1:

ry8806 commented:
hi, I've updated it to work with config (and updated the README and added more tests) - had a bit of trouble with Git so let me know if I've rebased correctly - hopefully it should all be there!

ealves-pt commented on src/angular-local-storage.js:
```diff
@@ -311,9 +313,23 @@ angular
               }
+              /* Providing the secure parameter always takes precedence over config 
+               * (allows developer to mix and match secure + non-secure) */
+              if (isDefined(secure) && secure !== null) {
```
Why not replace this nested `ifs` for:
```javascript
if (typeof secure === 'boolean' && secure) {
  cookieDomain += '; secure';
}
```

I think it makes the code more readable.

ry8806 replied:
yeh, good catch!
thanks

ealves-pt replied:
Any update regarding this? Will you do the change or should I merge it as is?

ry8806 replied:
If I take the code above, we lose the functionality of being able to explicitly override what config says.

At the moment - if config specifies `secure = true` I can call `cookie.set("key", "value", 3, false)` and it allows me to override what the config says (by going into the first "if" and obviously not going into the `if-else`).

If I take the code above and we specify secure as `false` we lose that functionality and will fall in to the `if-else` which will take the value from config. Which renders the secure parameter on `cookie.set` useless...

what do you think???

ealves-pt commented:
Nice PR. Just commented one thing and I will merge it after your feedback.

Thanks for the work :+1:

ry8806 commented:
I've just updated the type of checks on the if statements - they now check `typeof === 'boolean'` instead of the undefined check before - however my comment in reply to yours (above) still stands :smile: (as the flow hasn't changed)

ealves-pt commented on src/angular-local-storage.js:
```diff
@@ -318,9 +320,23 @@ angular
+                  }
+                  // else - secure has been supplied but isn't true - so don't set secure flag, regardless of what config says
+              }
+              else if (typeof secure === 'boolean' && cookie.secure === true) {
```
The `typeof secure === 'boolean'` should not be needed here. We reach this `else if` if `secure` is not a `boolean`. Besides this your explanation makes total sense. Thanks for the input :+1:

ry8806 replied:
no worries, thanks for helping getting this in!

ry8806 commented:
hi, just wondering if you've had a chance to look at getting this in yet?

grevory commented:
I see a syntax error on line 98 in localStorageSpec.js.

ry8806 commented:
I've corrected the syntax error - apologies!

grevory commented:
It looks like it is out of sync with master. Can you resolve this first? Thanks.

grevory commented:
I don't want your hard work to go unused. Think you could correct the conflict with master?

ry8806 commented:
yes, I can have a look and see what I can do!

ry8806 commented:
it seems quite far behind, what do you reckon, just pull from master and re-do my changes on top, or rebase???

jonkoops commented:
@ry8806 I would squash your commits into one and rebase from master.

ry8806 commented:
@jonkoops help would be much appreciated, only familiar with the (very) basic git stuff.
no idea how to squash and then rebase before all hell breaks loose on my machine! 😄

jonkoops commented:
@ry8806 It seems there are some commits in your pull request which are not part of the functionality itself. Is this correct?

jonkoops commented:
Looks like you pulled in some changes from the upstream master which are now part of your pull request. Could you give me access to your repository so I can correct it? I would love to explain all the inner workings of rebasing but I unfortunately do not have the time.

ry8806 commented:
think you should have access to the repo, no worries, thanks for helping!

i'll definitely have to read up more on the inner workings of squash/rebase when I find time

jonkoops commented:
@ry8806 I've squashed all the changes into a single commit, and I've rebased with the upstream master. We should be all good to merge now (if the tests run correctly).

ry8806 commented:
massive thanks @jonkoops

jonkoops commented:
Looks like some of the tests are failing, did I acidentally break something or were those tests failing before too?

ry8806 commented:
not sure, i'll check and fix 😄

jonkoops commented:
Ok, I've rebased your changes into the base commit. With the tests passing we should be good, for realsies this time.

jonkoops commented:
@grevory Any chance you could review and possibly merge?

jonkoops commented:
@grevory Bump.

![bump](https://cloud.githubusercontent.com/assets/695720/17726838/89fb9c3c-6455-11e6-83c1-9e4b2a8e34f4.gif)

grevory commented:
Sorry. Been travelling for work and pleasure. Thanks a ton!
